src/nbl/video/CVulkanPhysicalDevice.h

```diff
@@ -101,72 +100,163 @@ class CVulkanPhysicalDevice final : public IPhysicalDevice
         }
     }
             
-    inline VkPhysicalDevice getInternalObject() const { return m_vkphysdev; }
+    inline VkPhysicalDevice getInternalObject() const { return m_vkPhysicalDevice; }
             
     E_API_TYPE getAPIType() const override { return EAT_VULKAN; }
 
-    std::vector<ISurface::SFormat> getAvailableFormatsForSurface(const ISurface* surface) const override
+    // Todo(achal)
+    IDebugCallback* getDebugCallback() override { return nullptr; }
+
+    void getAvailableFormatsForSurface(const ISurface* surface, uint32_t& formatCount,
+        ISurface::SFormat* formats) const override
     {
-        const ISurfaceVK* vk_surface = static_cast<const ISurfaceVK*>(surface); // Todo(achal): This is problematic, if passed `surface` isn't a vulkan surface
+        constexpr uint32_t MAX_SURFACE_FORMAT_COUNT = 1000u;
+
+        if (surface->getAPIType() != EAT_VULKAN)
+            return;
 
-        uint32_t formatCount = 0u;
-        vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkphysdev, vk_surface->m_surface, &formatCount, nullptr);
-        std::vector<VkSurfaceFormatKHR> formats(formatCount);
-        vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkphysdev, vk_surface->m_surface, &formatCount, formats.data());
+        // Todo(achal): not sure yet, how would I handle multiple platforms without making
+        // this function templated
+        VkSurfaceKHR vk_surface = static_cast<const CSurfaceVulkanWin32*>(surface)->getInternalObject();
 
-        std::vector<ISurface::SFormat> result(formatCount);
+        VkResult retval = vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkPhysicalDevice, vk_surface,
+            &formatCount, nullptr);
+
+        // Todo(achal): Would there be a need to handle VK_INCOMPLETE separately?
+        if ((retval != VK_SUCCESS) && (retval != VK_INCOMPLETE))
+        {
+            formatCount = 0u;
+            return;
+        }
+
+        if (!formats)
+            return;
+
+        VkSurfaceFormatKHR vk_formats[MAX_SURFACE_FORMAT_COUNT];
```

Comment:
you could abuse `formats` as a temporary storage for `vk_formats`

```cpp
auto* vk_formats = reintepret_cast<VkSurfaceFormatKHR*>(formats);
```

the only thing you need to make sure of is `static_assert(sizeof(ISurface::SFormat)>=sizeof(VkSurfaceFormatKHR))`

Reply:
I also dont get why you call `vkGetPhysicalDeviceSurfaceFormatsKHR` twice